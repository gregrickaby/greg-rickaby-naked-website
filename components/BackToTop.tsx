'use client'

import {IconArrowUp} from '@tabler/icons-react'
import {useEffect, useState} from 'react'

/**
 * The back to top component.
 */
export default function BackToTop() {
  const [showButton, setShowButton] = useState(false)
  const buttonText = 'Go back to the top of the page'

  /**
   * Scroll handler.
   */
  const handleScroll = () => {
    window.scrollTo({
      top: 0,
      behavior: 'smooth'
    })
  }

  /**
   * Effect for showing the back to top button.
   */
  useEffect(() => {
    // Scroll event handler.
    const handleScroll = () => {
      setShowButton(window.scrollY > 200)
    }

    // Add event listener.
    window.addEventListener('scroll', handleScroll)

    // Cleanup event listener.
    return () => window.removeEventListener('scroll', handleScroll)
  }, [])

  return showButton ? (
    <button
      aria-label={buttonText}
      className="button fixed bottom-8 right-3 rounded-full bg-gray-400 p-2 text-white shadow-lg hover:bg-gray-500 lg:right-12 dark:bg-gray-800 hover:dark:bg-gray-700"
      data-testid="back-to-top"
      onClick={handleScroll}
      title={buttonText}
    >
      <IconArrowUp height="32" width="32" />
      <span className="sr-only">{buttonText}</span>
    </button>
  ) : null
}
