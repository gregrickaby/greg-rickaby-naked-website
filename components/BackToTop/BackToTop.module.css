.backToTop {
  @apply fixed bottom-8 right-3 lg:right-12;
}
