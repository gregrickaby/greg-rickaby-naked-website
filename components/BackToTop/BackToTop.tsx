'use client'

import {IconArrowUp} from '@tabler/icons-react'
import clsx from 'clsx'
import {useEffect, useState} from 'react'
import styles from './BackToTop.module.css'

/**
 * The back to top component.
 */
export function BackToTop() {
  const [showButton, setShowButton] = useState(false)
  const buttonText = 'Go back to the top of the page'

  /**
   * Scroll handler.
   */
  const handleScroll = () => {
    window.scrollTo({
      top: 0,
      behavior: 'smooth'
    })
  }

  /**
   * Effect for showing the back to top button.
   */
  useEffect(() => {
    // Scroll event handler.
    const handleScroll = () => {
      setShowButton(window.scrollY > 200)
    }

    // Add event listener.
    window.addEventListener('scroll', handleScroll)

    // Cleanup event listener.
    return () => window.removeEventListener('scroll', handleScroll)
  }, [])

  return showButton ? (
    <button
      aria-label={buttonText}
      className={clsx('button', styles.backToTop)}
      data-testid="back-to-top"
      onClick={handleScroll}
      title={buttonText}
    >
      <IconArrowUp height="32" width="32" />
      <span className="sr-only">{buttonText}</span>
    </button>
  ) : null
}
