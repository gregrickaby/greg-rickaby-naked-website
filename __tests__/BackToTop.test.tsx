import BackToTop from '@/components/BackToTop'
import '@testing-library/jest-dom'
import {act, fireEvent, render, screen, waitFor} from '@testing-library/react'
import {axe} from 'jest-axe'
import {describe, expect, vi} from 'vitest'

// Mock the IconArrowUp component to avoid rendering actual SVG during tests.
vi.mock('@tabler/icons-react', () => ({
  IconArrowUp: () => <div>Icon</div>
}))

// Scroll helper function.
const scrollTo = (y: number) => {
  Object.defineProperty(window, 'scrollY', {value: y, writable: true})
  window.dispatchEvent(new Event('scroll'))
}

describe('BackToTop Component', () => {
  beforeAll(() => {
    window.scrollTo = vi.fn() // Mock window.scrollTo method.
  })

  beforeEach(() => {
    window.scrollTo(0, 0) // Reset scroll position before each test.
  })

  it('should not render button when scroll position is less than 200', () => {
    render(<BackToTop />)
    expect(screen.queryByTestId('back-to-top')).not.toBeInTheDocument()
  })

  it('renders button when scroll position is greater than 200', async () => {
    render(<BackToTop />)

    act(() => {
      scrollTo(201)
    })

    await waitFor(() => {
      expect(screen.getByTestId('back-to-top')).toBeInTheDocument()
    })
  })

  it('scrolls to top when button is clicked', async () => {
    render(<BackToTop />)

    act(() => {
      scrollTo(201)
    })

    const button = await waitFor(() => screen.getByTestId('back-to-top'))

    act(() => {
      fireEvent.click(button)
    })

    await waitFor(() => {
      expect(window.scrollTo).toHaveBeenCalledWith({top: 0, behavior: 'smooth'})
    })
  })

  it('button has correct aria-label and title attributes', async () => {
    render(<BackToTop />)

    act(() => {
      scrollTo(201)
    })

    const button = await waitFor(() => screen.getByTestId('back-to-top'))

    expect(button).toHaveAttribute(
      'aria-label',
      'Go back to the top of the page'
    )
    expect(button).toHaveAttribute('title', 'Go back to the top of the page')
  })

  it('should not have any accessibility issues', async () => {
    render(<BackToTop />)

    act(() => {
      scrollTo(201)
    })

    const button = await waitFor(() => screen.getByTestId('back-to-top'))
    const results = await axe(button)
    expect(results).toHaveNoViolations()
  })

  it('should match snapshot', async () => {
    const {container} = render(<BackToTop />)

    act(() => {
      scrollTo(201)
    })

    await waitFor(() => screen.getByTestId('back-to-top'))
    expect(container).toMatchSnapshot()
  })
})
