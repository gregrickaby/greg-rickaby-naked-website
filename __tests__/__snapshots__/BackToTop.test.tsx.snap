// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BackToTop Component > should match snapshot 1`] = `
<div>
  <button
    aria-label="Go back to the top of the page"
    class="button fixed bottom-8 right-3 rounded-full bg-gray-400 p-2 text-white shadow-lg hover:bg-gray-500 lg:right-12 dark:bg-gray-800 hover:dark:bg-gray-700"
    data-testid="back-to-top"
    title="Go back to the top of the page"
  >
    <div>
      Icon
    </div>
    <span
      class="sr-only"
    >
      Go back to the top of the page
    </span>
  </button>
</div>
`;
